langchain/chains/openai_functions/citation_fuzzy_match.py

```diff
@@ -0,0 +1,82 @@
+from typing import List
+from pydantic import Field, BaseModel
+
+from langchain.base_language import BaseLanguageModel
+from langchain.output_parsers.openai_functions import (
+    PydanticOutputFunctionsParser,
+)
+from langchain.chains.llm import LLMChain
+from langchain.prompts.chat import ChatPromptTemplate, HumanMessagePromptTemplate
+from langchain.schema import SystemMessage, HumanMessage
+
+
+class Fact(BaseModel):
+    """
+    Class representing single statement.
+    Each fact has a body and a list of sources.
+    If there are multiple facts make sure to break them apart such that each one only uses a set of sources that are relevant to it.
+    """
+
+    fact: str = Field(..., description="Body of the sentence, as part of a response")
+    substring_quote: List[str] = Field(
+        ...,
+        description="Each source should be a direct quote from the context, as a substring of the original content",
+    )
+
+    def _get_span(self, quote, context, errs=100):
+        import regex
+
+        minor = quote
+        major = context
+
+        errs_ = 0
+        s = regex.search(f"({minor}){{e<={errs_}}}", major)
```

Comment:
If you're using the same regex for many times, then it's better to compile and use it. It should give a performance boost.

```
match_regx = re.compile(f"({minor}){{e<={errs_}}}")
s = match_regx.search(major)
while s is None and errs_ <= errs:
            errs_ += 1
            s = match_regx.search(major)
```